File: backend/docker_python/sandbox.py
import sys
import resource
import pyseccomp as seccomp


MEMORY_LIMIT = 64 * 1024 * 1024  # 64kb
CPU_TIME_LIMIT = 3  # 1sec
WRITE_LIMIT = 512  # 512bytes


def drop_user_permissions():
    # user can't access sensitive files (ex: Dockerfile)
    filter = seccomp.SyscallFilter(seccomp.ERRNO(seccomp.errno.EPERM))
    filter.add_rule(
        seccomp.ALLOW, "write", seccomp.Arg(0, seccomp.EQ, sys.stdout.fileno())
    )
    filter.add_rule(
        seccomp.ALLOW, "write", seccomp.Arg(0, seccomp.EQ, sys.stderr.fileno())
    )
    filter.load()


def set_mememory_limit():
    # v memory
    resource.setrlimit(resource.RLIMIT_AS, (MEMORY_LIMIT, MEMORY_LIMIT))
    # cpu timelimit
    resource.setrlimit(resource.RLIMIT_CPU, (CPU_TIME_LIMIT, CPU_TIME_LIMIT))
    # user code size limit 
    resource.setrlimit(resource.RLIMIT_FSIZE, (WRITE_LIMIT, WRITE_LIMIT))

def execute_python_code(code):
    set_mememory_limit()
    drop_user_permissions()
    exec(code)


if __name__ == "__main__":
    code = sys.argv[1]
    execute_python_code(code)

File: backend/docker_go/sandbox.go
package main

import (
	"fmt"
	"os"
	"syscall"
)

const (
	memoryLimit = 64 * 1024 * 1024 // 64MB
	cpuTimeLimit = 3 // 3 seconds
	writeLimit = 512 // 512 bytes
)

func setLimits() {
	// Set memory limit
	if err := syscall.Setrlimit(syscall.RLIMIT_AS, &syscall.Rlimit{Cur: memoryLimit, Max: memoryLimit}); err != nil {
		fmt.Println("Failed to set memory limit:", err)
		os.Exit(1)
	}

	// Set CPU time limit
	if err := syscall.Setrlimit(syscall.RLIMIT_CPU, &syscall.Rlimit{Cur: cpuTimeLimit, Max: cpuTimeLimit}); err != nil {
		fmt.Println("Failed to set CPU time limit:", err)
		os.Exit(1)
	}

	// Set file size (write) limit
	if err := syscall.Setrlimit(syscall.RLIMIT_FSIZE, &syscall.Rlimit{Cur: writeLimit, Max: writeLimit}); err != nil {
		fmt.Println("Failed to set file size limit:", err)
		os.Exit(1)
	}
}

func main() {
	setLimits()

	code := os.Getenv("GO_CODE")
	if code == "" {
		fmt.Println("No code provided")
		os.Exit(1)
	}

	fmt.Println("Executing provided code:")
	fmt.Println(code)
	executeProvidedCode(code)
}

func executeProvidedCode(code string) {
	fmt.Println("Executing code...")
	fmt.Println(code)
}
